fortanix-vme/fortanix-vme-runner/src/lib.rs

```diff
@@ -286,60 +389,55 @@ impl Server {
+
                 Ok(())
             },
             Err(e) => Err(e),
         }
     }
 
-    fn proxy_connection(remote: (&mut TcpStream, &str), proxy: (&mut VsockStream, &str)) {
+    fn proxy_connections(server: Arc<Server>) {
+        let mut closed_connections = Vec::new();
+
         loop {
-            let mut read_set = FdSet::new();
-            read_set.insert(remote.0.as_raw_fd());
-            read_set.insert(proxy.0.as_raw_fd());
-
-            if let Ok(_num) = select(None, Some(&mut read_set), None, None, None) {
-                if read_set.contains(remote.0.as_raw_fd()) {
-                    match Self::transfer_data(remote.0, remote.1, proxy.0, proxy.1) {
-                        Ok(0)  => {
-                            // According to the `Read` threat documentation, reading 0 bytes
-                            // indicates that the connection has been shutdown correctly. So we
-                            // close the proxy service
-                            // https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read
-                            break
-                        },
-                        Ok(_)  => (),
-                        Err(e) => {
-                            eprintln!("transfer from remote failed: {:?}", e);
-                            break;
-                        }
+            // Exchange messages on every proxy connection
+            // TODO: Store connections as a linked hash map so we don't need to keep a read lock
+            // over the HashMap while every connection is serviced
+            if let Ok(connections) = server.connections.read() {
+                for (key, connection) in connections.iter() {
+                    match connection.try_lock() {
+                        Ok(mut connection) => if !connection.proxy() {
+                                            connection.close();
+                                            closed_connections.push(key.clone());
+                                          }
+                        Err(_) => (),
                     }
                 }
-                if read_set.contains(proxy.0.as_raw_fd()) {
-                    match Self::transfer_data(proxy.0, proxy.1, remote.0, remote.1) {
-                        Ok(0)  => break,
-                        Ok(_)  => (),
-                        Err(e) => {
-                            eprintln!("transfer from proxy failed: {:?}", e);
-                            break;
-                        }
-                    }
+            }
+
+            // Remove closed connections
+            let mut num_connections = None;
+            if let Ok(mut connections) = server.connections.try_write() {
+                while let Some(k) = closed_connections.pop() {
+                    connections.remove(&k);
                 }
+                num_connections = Some(connections.len());
+            }
+
+            if num_connections == Some(0) {
+                thread::yield_now();
```

Comment:
Is there a risk of having excessive CPU usage when there are no connections? We basically keep polling in that case. An alternative would be to use a condition variable to block until we have connections again.